### DISCUSSION TITLE
Getting matching files list in script

### DISCUSSION THREAD

**Discussion by avudnez** [evilmartians/lefthook#578] [Q&A] [answered]
Hi,

I'm considering switching from pre-commit, which (when applicable) passes all files matching the glob for a defined script as input arguments.
It seems lefthook does not do that, and only has the template system with `{files}` for commands, which is unfortunate because I have a lot of bash scripts that could very well benefit from the same feature ?
Since lefthook encourages defining a file glob in the configuration, I don't want to have to duplicate it inside my script to get the same information that should be available.

**Comment by mrexox** [comment]
Hey @avudnez! If I correctly understood you need to get all files in your repo and filter them via some glob. For that you can specify `files` command like `git ls-tree --full-tree -r --name-only HEAD` then provide a `glob` parameter to filter them.

If it's not what you want, please, describe your case in more details, maybe with an example repo.

**Reply by avudnez** [reply]
Here's the script I'm using with pre-commit at the moment:
```bash
if git rev-parse --verify HEAD >/dev/null 2>&1
then
	against=HEAD
else
	# Initial commit: diff against an empty tree object
	against=$(git hash-object -t tree /dev/null)
fi

# Redirect output to stderr.
exec 1>&2

## 	Helpers
##  -------

HOOK_EXITCODE=0

function die()
{
	>&2 echo " (!!) ERROR: $*"
	HOOK_EXITCODE=2
}

FILES_TO_CHECK=$@

if [ -z "$FILES_TO_CHECK" ]; then
    exit $HOOK_EXITCODE # no files to check, bail
fi

function diff-lines()
{
    local path=
    local line=
    while read; do
        esc=$'\033'
        if [[ $REPLY =~ ---\ (a/)?.* ]]; then
            continue
        elif [[ $REPLY =~ \+\+\+\ (b/)?([^[:blank:]$esc]+).* ]]; then
            path=${BASH_REMATCH[2]}
        elif [[ $REPLY =~ @@\ -[0-9]+(,[0-9]+)?\ \+([0-9]+)(,[0-9]+)?\ @@.* ]]; then
            line=${BASH_REMATCH[2]}
        elif [[ $REPLY =~ ^($esc\[[0-9;]+m)*\+ ]]; then
            echo "$path:$line:$REPLY"
            if [[ ${BASH_REMATCH[2]} != - ]]; then
                ((line++))
            fi
        fi
    done
}

# Print the offending lines and fail if %NOCOMMIT% tags are present
# -> only check white listed files
if git diff --cached $against -- $FILES_TO_CHECK | diff-lines | grep --color=auto '%NOCOMMIT%'; then
	die "Found %NOCOMMIT% tag(s)"
fi

## 	Finally return an exit code indicating whether any check failed or not
##  ----------------------------------------------------------------------

if [ $HOOK_EXITCODE -ne 0 ]; then
	exit $HOOK_EXITCODE
fi
```
The hook configuration is:
```yaml
    -   id: check-nocommit
        name: Check for %NOCOMMIT% tag
        language: script
        types_or: [asciidoc, c, c++, c#, markdown, python, shell]
        exclude: |
            (?x)^(
                README.md|
                README.md|
                pre-commit.sh|
                check_nocommit.sh
            )$
```
As you can see
- the hook is only executed when some source types are being commited
- except special cases
- the script receives a file list via its arguments from the pre-commit runner and forwards it directly to git

I'm only interested in checking files that match the `types_or` parameter, which would be `glob` in lefthook AFAICT.
However, with lefthook I would have to duplicate the glob inside my script to only look at the wanted files. There could be other files staged that I don't care about.

**Reply by mrexox** [reply, marked as answer]
Ok, I think I know what may fit your case. You need to use `commands` instead of `scripts` and pass the file list to the script manually, like this:

```yml
# lefthook.yml

pre-commit:
  commands:
    check-nocommit:
      glob: "*.{adoc,c,cpp,cs,md,py,sh}"
      run: scripts/check-nocommit "{staged_files}"
```

And in the script you need to iterate the args.

Does it sound helpful?

**Reply by avudnez** [reply]
Thanks, that does work. However, to do this on windows I need to use `run: bash path/to/check-nocommit.sh "{staged_files}"`.

**Reply by mj3c** [reply]
Hey @mrexox, would something like this still be possible when using a `remotes` lefthook config?

Basically run a script as `commands` and pass `{staged_files}`, but how do I know the script path if lefthook config is stored remotely? Would I have to hardcode something like `.git/info/lefthook-remotes/<repo>-<branch>/.lefthook/pre-commit/myscript.sh`, or is there a better way?

**Reply by mrexox** [reply]
I think this may not work well with the remotes feature. I can suggest moving the script content into a command (if you want to utilize {staged_files} template), like this:

```yml
pre-commit:
  commands:
    command:
      run: |
        for file in {stage_files}; do
        ...
        done
```

using `.git/info/lefthook-remotes/...` paths in unreliable
